suggestions.go

```diff
@@ -0,0 +1,75 @@
+package cli
+
+import (
+	"fmt"
+
+	"github.com/antzucaro/matchr"
+)
+
+const didYouMeanTemplate = "Did you mean '%s'?"
+
+func (a *App) suggestFlagFromError(err error, command string) (string, error) {
+	flag, parseErr := flagFromError(err)
+	if parseErr != nil {
+		return "", err
+	}
+
+	flags := a.Flags
+	if command != "" {
+		cmd := a.Command(command)
+		if cmd == nil {
+			return "", err
+		}
+		flags = cmd.Flags
+	}
+
+	suggestion := a.suggestFlag(flags, flag)
+	if len(suggestion) == 0 {
+		return "", err
+	}
+
+	return fmt.Sprintf(didYouMeanTemplate+"\n\n", suggestion), nil
+}
+
+func (a *App) suggestFlag(flags []Flag, provided string) (suggestion string) {
+	distance := 0.0
+
+	for _, flag := range flags {
+		flagNames := flag.Names()
+		if !a.HideHelp {
+			flagNames = append(flagNames, HelpFlag.Names()...)
+		}
+		for _, name := range flagNames {
+			newDistance := matchr.JaroWinkler(name, provided, true)
+			if newDistance > distance {
+				distance = newDistance
+				suggestion = name
+			}
+		}
+	}
+
+	if len(suggestion) == 1 {
+		suggestion = "-" + suggestion
+	} else if len(suggestion) > 1 {
+		suggestion = "--" + suggestion
+	}
+
+	return suggestion
+}
+
+// suggestCommand takes a list of commands and a provided string to suggest a
+// command name
+func suggestCommand(commands []*Command, provided string) (suggestion string) {
+	distance := 0.0
+	for _, command := range commands {
```

Comment:
⭐ _(nice to have)_ I'm not sure about this suggestion, but: would it be a good idea to consider suggesting sub commands too? I'm imagining an example CLI that has exactly these 3 commands

```
docker info
docker info containers
docker info images
```

and I input this

```
docker containers
```

I would be expecting to get a suggestion like

```
did you mean: docker info containers?
```

whereas I believe this code with give me

```
did you mean: docker info?
```

is that a correct assumption?

Reply:
Yes it is correct that it would only suggest `docker info` in that case, because can only suggest one command with the current implementation. I'm also not sure if we should add such a _traversal suggestion_ to the users. It might be helpful but it also might result in confusing them. 😁

Reply:
^^ fair!